File: lib/roster.ts
import type { Player, Position, TeamRosterSeed } from './types';

// Pure roster queries — no dependency on the team registry, so client components
// can import these (and resolveUnit) without bundling every team's data. They only
// read players, so they take the lighter seed shape (works for full rosters too).

export function getPlayerById(roster: TeamRosterSeed, id: string): Player | undefined {
  return roster.players.find((p) => p.id === id);
}

// Deterministic order: depthRank first, then jersey number as a stable tiebreak.
// Multiple players share a depthRank at a position (e.g. 3 WR1s), so the tiebreak
// is what makes field slot assignment (WR index 0/1/2) reproducible across edits.
// Shared by getPlayersByPosition below and lib/formations.ts's
// getPlayersByPositionGroup, which filters by broad group instead of exact position
// but needs the same depth order.
export function byDepthOrder(a: Player, b: Player): number {
  return a.depthRank - b.depthRank || (a.order ?? a.number) - (b.order ?? b.number);
}

export function getPlayersByPosition(roster: TeamRosterSeed, position: Position): Player[] {
  return roster.players.filter((p) => p.position === position).sort(byDepthOrder);
}

File: ios/Depth/Domain/Roster.swift
import Foundation

// Mirrors lib/utils/roster/roster.ts — pure roster queries with no dependency on team
// identity, so they work from just a Roster (players + specialTeams).

func getPlayer(in roster: Roster, id: String) -> Player? {
    roster.players.first { $0.id == id }
}

/// Deterministic order: depthRank first, then jersey number as a stable tiebreak
/// (or `order`, when a user override set one). Mirrors byDepthOrder exactly, including
/// operator precedence: the tiebreak only runs when depthRank ties.
func byDepthOrder(_ a: Player, _ b: Player) -> Bool {
    if a.depthRank != b.depthRank {
        return a.depthRank < b.depthRank
    }
    return (a.order ?? a.number) < (b.order ?? b.number)
}

func getPlayers(in roster: Roster, at position: Position) -> [Player] {
    roster.players.filter { $0.position == position }.sorted(by: byDepthOrder)
}
